pkg/trafficmngr/nftables/nftables.go

```diff
@@ -0,0 +1,317 @@
+// Copyright 2024 flannel authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//go:build !windows
+// +build !windows
+
+package nftables
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	log "k8s.io/klog/v2"
+
+	"github.com/flannel-io/flannel/pkg/ip"
+	"github.com/flannel-io/flannel/pkg/lease"
+	"sigs.k8s.io/knftables"
+)
+
+const (
+	ipv4Table    = "flannel-ipv4"
+	ipv6Table    = "flannel-ipv6"
+	forwardChain = "forward"
+	postrtgChain = "postrtg"
+	//maximum delay in second to clean-up when the context is cancelled
+	cleanUpDeadline = 15
+)
+
+type NFTablesManager struct {
+	nftv4 knftables.Interface
+	nftv6 knftables.Interface
+}
+
+func (nftm *NFTablesManager) Init(ctx context.Context, wg *sync.WaitGroup) error {
+	var err error
+	nftm.nftv4, err = initTable(ctx, knftables.IPv4Family, ipv4Table)
+	if err != nil {
+		return err
+	}
+	nftm.nftv6, err = initTable(ctx, knftables.IPv6Family, ipv6Table)
+	if err != nil {
+		return err
+	}
+
+	wg.Add(1)
+	go func() {
+		<-ctx.Done()
+		log.Info("Cleaning-up flannel tables...")
+
+		cleanupCtx, cleanUpCancelFunc := context.WithTimeout(context.Background(), cleanUpDeadline*time.Second)
+		defer cleanUpCancelFunc()
+		err := nftm.cleanUp(cleanupCtx)
+		log.Errorf("nftables: error while cleaning-up: %v", err)
+		wg.Done()
+	}()
+	return nil
+}
+
+// create a new table and returns the interface to interact with it
+func initTable(ctx context.Context, ipFamily knftables.Family, name string) (knftables.Interface, error) {
+	nft, err := knftables.New(ipFamily, name)
+	if err != nil {
+		return nil, fmt.Errorf("no nftables support: %v", err)
+	}
+	tx := nft.NewTransaction()
+
+	tx.Add(&knftables.Table{
+		Comment: knftables.PtrTo("rules for " + name),
+	})
+	err = nft.Run(ctx, tx)
+	if err != nil {
+		return nil, fmt.Errorf("nftables: couldn't initialise table %s: %v", name, err)
+	}
+	return nft, nil
+}
+
+// It is needed when using nftables? accept seems to be the default
+// warning: never add a default 'drop' policy on the forwarChain as it breaks connectivity to the node
+func (nftm *NFTablesManager) SetupAndEnsureForwardRules(ctx context.Context,
+	flannelIPv4Network ip.IP4Net, flannelIPv6Network ip.IP6Net, resyncPeriod int) {
+	if !flannelIPv4Network.Empty() {
+		log.Infof("Changing default FORWARD chain policy to ACCEPT")
+		tx := nftm.nftv4.NewTransaction()
+
+		//TODO how to express that the default is drop?
```

Comment:
Do we need to DROP by default? I think we can add a simple "drop" rule at the end of the chain. I saw that on the class the `policy` value is missing it should be something like this `type filter hook forward priority -100; policy drop;`

Reply:
Yes I was confused by the question. It shouldn't drop but it should continue the chain in case it's not matching the flannel subnet using the default configuration of the node.

Reply:
I think that's the default behavior so it should be fine as it is.
From the comment in flannel code, the forward rules seem needed for old versions of docker anyway.